// peer-calls/lib/client/reducers/media.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
var constants_1 = require("../constants");
var defaultState = {
    devices: [],
    video: { facingMode: 'user' },
    audio: true,
    dialState: constants_1.DIAL_STATE_HUNG_UP,
    loading: false,
    error: '',
    autoplayError: false
};
function handleEnumerate(state, action) {
    switch (action.status) {
        case 'resolved':
            return __assign(__assign({}, state), { loading: false, devices: action.payload });
        case 'pending':
            return __assign(__assign({}, state), { loading: true });
        case 'rejected':
            return __assign(__assign({}, state), { loading: false, error: 'Could not retrieve media devices' });
    }
}
exports.handleEnumerate = handleEnumerate;
function handleMediaStream(state, action) {
    switch (action.status) {
        case 'resolved':
            return __assign({}, state);
        case 'rejected':
            return __assign(__assign({}, state), { error: action.payload.message });
        default:
            return state;
    }
}
exports.handleMediaStream = handleMediaStream;
function handlePlay(state, action) {
    switch (action.status) {
        case 'pending':
        case 'resolved':
            return __assign(__assign({}, state), { autoplayError: false });
        case 'rejected':
            console.log('play rejected', action.payload.name);
            if (action.payload.name !== 'NotAllowedError') {
                return state;
            }
            return __assign(__assign({}, state), { autoplayError: true });
        default:
            return state;
    }
}
exports.handlePlay = handlePlay;
function handleDial(state, action) {
    switch (action.status) {
        case 'pending':
            return __assign(__assign({}, state), { dialState: constants_1.DIAL_STATE_DIALLING });
        case 'resolved':
            return __assign(__assign({}, state), { dialState: constants_1.DIAL_STATE_IN_CALL });
        case 'rejected':
            return __assign(__assign({}, state), { dialState: constants_1.DIAL_STATE_HUNG_UP });
        default:
            return state;
    }
}
exports.handleDial = handleDial;
function media(state, action) {
    if (state === void 0) { state = defaultState; }
    switch (action.type) {
        case constants_1.MEDIA_ENUMERATE:
            return handleEnumerate(state, action);
        case constants_1.MEDIA_AUDIO_CONSTRAINT_SET:
            return __assign(__assign({}, state), { audio: action.payload });
        case constants_1.MEDIA_VIDEO_CONSTRAINT_SET:
            return __assign(__assign({}, state), { video: action.payload });
        case constants_1.MEDIA_STREAM:
            return handleMediaStream(state, action);
        case constants_1.MEDIA_PLAY:
            return handlePlay(state, action);
        case constants_1.DIAL:
            return handleDial(state, action);
        case constants_1.HANG_UP:
            return __assign(__assign({}, state), { dialState: constants_1.DIAL_STATE_HUNG_UP });
        default:
            return state;
    }
}
exports["default"] = media;
