// peer-calls/lib/client/reducers/messages.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var constants = __importStar(require("../constants"));
var defaultState = {
    list: [],
    count: 0
};
function convertNotificationToMessage(action) {
    return {
        userId: '[PeerCalls]',
        message: action.payload.message,
        system: true,
        timestamp: new Date().toLocaleString()
    };
}
function messages(state, action) {
    if (state === void 0) { state = defaultState; }
    switch (action.type) {
        case constants.NOTIFY:
            return __assign(__assign({}, state), { list: __spreadArrays(state.list, [convertNotificationToMessage(action)]) });
        case constants.MESSAGE_ADD:
            return {
                count: action.payload.system ? state.count : state.count + 1,
                list: __spreadArrays(state.list, [action.payload])
            };
        default:
            return state;
    }
}
exports["default"] = messages;
