// peer-calls/lib/client/reducers/peers.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var debug_1 = __importDefault(require("debug"));
var forEach_1 = __importDefault(require("lodash/forEach"));
var omit_1 = __importDefault(require("lodash/omit"));
var constants = __importStar(require("../constants"));
var debug = debug_1["default"]('peercalls');
var defaultState = {};
var localStreams = {
    camera: undefined,
    desktop: undefined
};
function removeTrackFromPeer(peer, track, stream) {
    try {
        peer.removeTrack(track, stream);
    }
    catch (err) {
        debug('peer.removeTrack: %s', err);
    }
}
function handleRemoveStream(state, action) {
    var stream = action.payload.stream;
    if (action.payload.userId === constants.ME) {
        forEach_1["default"](state, function (peer) {
            stream.getTracks().forEach(function (track) {
                removeTrackFromPeer(peer, track, stream);
            });
        });
    }
    return state;
}
function handleMediaStream(state, action) {
    if (action.status !== 'resolved') {
        return state;
    }
    var streamType = action.payload.type;
    if (action.payload.userId === constants.ME &&
        streamType) {
        forEach_1["default"](state, function (peer) {
            var localStream = localStreams[streamType];
            localStream && localStream.getTracks().forEach(function (track) {
                removeTrackFromPeer(peer, track, localStream);
            });
            var stream = action.payload.stream;
            stream.getTracks().forEach(function (track) {
                peer.addTrack(track, stream);
            });
        });
        localStreams[streamType] = action.payload.stream;
    }
    return state;
}
function peers(state, action) {
    var _a;
    if (state === void 0) { state = defaultState; }
    switch (action.type) {
        case constants.PEER_ADD:
            return __assign(__assign({}, state), (_a = {}, _a[action.payload.userId] = action.payload.peer, _a));
        case constants.PEER_REMOVE:
            return omit_1["default"](state, [action.payload.userId]);
        case constants.HANG_UP:
            localStreams = {
                camera: undefined,
                desktop: undefined
            };
            setTimeout(function () {
                forEach_1["default"](state, function (peer) { return peer.destroy(); });
            });
            return defaultState;
        case constants.STREAM_REMOVE:
            return handleRemoveStream(state, action);
        case constants.MEDIA_STREAM:
            return handleMediaStream(state, action);
        default:
            return state;
    }
}
exports["default"] = peers;
