// peer-calls/lib/client/actions/CallActions.test.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
jest.mock('../socket');
jest.mock('../window');
jest.mock('./SocketActions');
var CallActions = __importStar(require("./CallActions"));
var SocketActions = __importStar(require("./SocketActions"));
var constants = __importStar(require("../constants"));
var socket_1 = __importDefault(require("../socket"));
var redux_1 = require("redux");
var reducers_1 = __importDefault(require("../reducers"));
var middlewares_1 = require("../middlewares");
jest.useFakeTimers();
describe('CallActions', function () {
    var callActions;
    function allActions(state, action) {
        if (state === void 0) { state = []; }
        return __spreadArrays(state, [action]);
    }
    var configureStore = function () { return redux_1.createStore(redux_1.combineReducers(__assign(__assign({}, reducers_1["default"]), { allActions: allActions })), redux_1.applyMiddleware.apply(void 0, middlewares_1.middlewares)); };
    var store;
    beforeEach(function () {
        store = redux_1.createStore(redux_1.combineReducers({ allActions: allActions }), redux_1.applyMiddleware.apply(void 0, middlewares_1.middlewares));
        callActions = redux_1.bindActionCreators(CallActions, store.dispatch);
        SocketActions.handshake.mockReturnValue(jest.fn());
    });
    afterEach(function () {
        jest.runAllTimers();
        socket_1["default"].removeAllListeners();
    });
    describe('init', function () {
        it('dispatches init action when connected', function () { return __awaiter(void 0, void 0, void 0, function () {
            var promise;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promise = callActions.init();
                        socket_1["default"].emit('connect', undefined);
                        return [4 /*yield*/, promise];
                    case 1:
                        _a.sent();
                        expect(store.getState().allActions.slice(1)).toEqual([{
                                type: constants.NOTIFY,
                                payload: {
                                    id: jasmine.any(String),
                                    message: 'Connected to server socket',
                                    type: 'warning'
                                }
                            }, {
                                type: constants.INIT
                            }]);
                        return [2 /*return*/];
                }
            });
        }); });
        it('calls dispatches disconnect message on disconnect', function () { return __awaiter(void 0, void 0, void 0, function () {
            var promise;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promise = callActions.init();
                        socket_1["default"].emit('connect', undefined);
                        socket_1["default"].emit('disconnect', undefined);
                        expect(store.getState().allActions.slice(1)).toEqual([{
                                type: constants.NOTIFY,
                                payload: {
                                    id: jasmine.any(String),
                                    message: 'Connected to server socket',
                                    type: 'warning'
                                }
                            }, {
                                type: constants.INIT
                            }, {
                                type: constants.NOTIFY,
                                payload: {
                                    id: jasmine.any(String),
                                    message: 'Server socket disconnected',
                                    type: 'error'
                                }
                            }]);
                        return [4 /*yield*/, promise];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
        it('dispatches alert when failed to get media stream', function () { return __awaiter(void 0, void 0, void 0, function () {
            var promise;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promise = callActions.init();
                        socket_1["default"].emit('connect', undefined);
                        return [4 /*yield*/, promise];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); });
    });
});
