// peer-calls/lib/client/actions/NotifyActions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var uniqueId_1 = __importDefault(require("lodash/uniqueId"));
var constants = __importStar(require("../constants"));
function format(string, args) {
    string = args
        .reduce(function (string, arg, i) { return string.replace('{' + i + '}', arg); }, string);
    return string;
}
function notify(type, args) {
    var string = args[0] || '';
    var message = format(string, Array.prototype.slice.call(args, 1));
    var id = uniqueId_1["default"]('notification');
    var payload = { id: id, type: type, message: message };
    return addNotification(payload);
}
exports.info = function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return notify('info', args);
};
exports.warning = function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return notify('warning', args);
};
exports.error = function () {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return notify('error', args);
};
function addNotification(payload) {
    return {
        type: constants.NOTIFY,
        payload: payload
    };
}
function dismissNotification(id) {
    return {
        type: constants.NOTIFY_DISMISS,
        payload: { id: id }
    };
}
exports.dismissNotification = dismissNotification;
exports.clear = function () { return ({
    type: constants.NOTIFY_CLEAR
}); };
