// peer-calls/lib/client/actions/SocketActions.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var NotifyActions = __importStar(require("../actions/NotifyActions"));
var PeerActions = __importStar(require("../actions/PeerActions"));
var constants = __importStar(require("../constants"));
var debug_1 = __importDefault(require("debug"));
var debug = debug_1["default"]('peercalls');
var SocketHandler = /** @class */ (function () {
    function SocketHandler(options) {
        var _this = this;
        this.handleSignal = function (_a) {
            var userId = _a.userId, signal = _a.signal;
            var getState = _this.getState;
            var peer = getState().peers[userId];
            // debug('socket signal, userId: %s, signal: %o', userId, signal);
            if (!peer)
                return debug('user: %s, no peer found', userId);
            peer.signal(signal);
        };
        this.handleUsers = function (_a) {
            var initiator = _a.initiator, users = _a.users;
            var _b = _this, socket = _b.socket, stream = _b.stream, dispatch = _b.dispatch, getState = _b.getState;
            debug('socket users: %o', users);
            _this.dispatch(NotifyActions.info('Connected users: {0}', users.length));
            var peers = _this.getState().peers;
            debug('active peers: %o', Object.keys(peers));
            users
                .filter(function (user) {
                return user.userId && !peers[user.userId] && user.userId !== _this.userId;
            })
                .forEach(function (user) { return PeerActions.createPeer({
                socket: socket,
                user: {
                    // users without id should be filtered out
                    id: user.userId
                },
                initiator: initiator,
                stream: stream
            })(dispatch, getState); });
        };
        this.socket = options.socket;
        this.roomName = options.roomName;
        this.stream = options.stream;
        this.dispatch = options.dispatch;
        this.getState = options.getState;
        this.userId = options.userId;
    }
    return SocketHandler;
}());
function handshake(options) {
    var socket = options.socket, roomName = options.roomName, stream = options.stream, userId = options.userId, store = options.store;
    var handler = new SocketHandler({
        socket: socket,
        roomName: roomName,
        stream: stream,
        dispatch: store.dispatch,
        getState: store.getState,
        userId: userId
    });
    // remove listeneres to make socket reusable
    removeEventListeners(socket);
    socket.on(constants.SOCKET_EVENT_SIGNAL, handler.handleSignal);
    socket.on(constants.SOCKET_EVENT_USERS, handler.handleUsers);
    debug('userId: %s', userId);
    socket.emit(constants.SOCKET_EVENT_READY, {
        room: roomName,
        userId: userId
    });
}
exports.handshake = handshake;
function removeEventListeners(socket) {
    var ee = socket;
    (ee.removeAllListeners)(constants.SOCKET_EVENT_SIGNAL);
    (ee.removeAllListeners)(constants.SOCKET_EVENT_USERS);
}
exports.removeEventListeners = removeEventListeners;
