// peer-calls/lib/client/containers/App.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var react_redux_1 = require("react-redux");
var CallActions_1 = require("../actions/CallActions");
var MediaActions_1 = require("../actions/MediaActions");
var NotifyActions_1 = require("../actions/NotifyActions");
var PeerActions_1 = require("../actions/PeerActions");
var StreamActions_1 = require("../actions/StreamActions");
var App_1 = __importDefault(require("../components/App"));
function mapStateToProps(state) {
    return {
        dialState: state.media.dialState,
        streams: state.streams,
        peers: state.peers,
        notifications: state.notifications,
        nicknames: state.nicknames,
        messages: state.messages.list,
        messagesCount: state.messages.count,
        windowStates: state.windowStates
    };
}
var mapDispatchToProps = {
    hangUp: CallActions_1.hangUp,
    minimizeToggle: StreamActions_1.minimizeToggle,
    sendMessage: PeerActions_1.sendMessage,
    dismissNotification: NotifyActions_1.dismissNotification,
    getDesktopStream: MediaActions_1.getDesktopStream,
    removeStream: StreamActions_1.removeStream,
    init: CallActions_1.init,
    onSendFile: PeerActions_1.sendFile,
    play: MediaActions_1.play
};
exports["default"] = react_redux_1.connect(mapStateToProps, mapDispatchToProps)(App_1["default"]);
