// peer-calls/lib/client/__mocks__/window.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.createObjectURL = jest.fn()
    .mockImplementation(function (object) { return 'blob://' + String(object); });
exports.revokeObjectURL = jest.fn();
var count = 0;
var MediaStream = /** @class */ (function () {
    function MediaStream() {
        this.addTrack = jest.fn();
        this.removeTrack = jest.fn();
        this.getTracks = jest.fn().mockReturnValue([{
                stop: jest.fn()
            }, {
                stop: jest.fn()
            }]);
        this.getVideoTracks = jest.fn().mockReturnValue([{
                enabled: true,
                stop: jest.fn()
            }]);
        this.getAudioTracks = jest.fn().mockReturnValue([{
                stop: jest.fn(),
                enabled: true
            }]);
        this.id = String(++count);
    }
    return MediaStream;
}());
exports.MediaStream = MediaStream;
var MediaStreamTrack = /** @class */ (function () {
    function MediaStreamTrack() {
    }
    return MediaStreamTrack;
}());
exports.MediaStreamTrack = MediaStreamTrack;
exports.navigator = window.navigator;
window.navigator.mediaDevices = {};
window.navigator.mediaDevices.enumerateDevices = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/, []];
    });
}); };
window.navigator.mediaDevices.getUserMedia = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/, {}];
    });
}); };
window.navigator.mediaDevices.getDisplayMedia = function () { return __awaiter(void 0, void 0, void 0, function () {
    return __generator(this, function (_a) {
        return [2 /*return*/, {}];
    });
}); };
exports.play = jest.fn();
exports.valueOf = jest.fn();
exports.callId = 'call1234';
exports.userId = 'user1234';
exports.nickname = 'nick1234';
exports.iceServers = [];
