// peer-calls/lib/client/components/Nickname.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var react_1 = __importDefault(require("react"));
var window_1 = require("../window");
var Nickname = /** @class */ (function (_super) {
    __extends(Nickname, _super);
    function Nickname() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Nickname.prototype.render = function () {
        if (this.props.localUser) {
            return (react_1["default"].createElement(MemoEditableNickname, { value: this.props.value, onChange: this.props.onChange }));
        }
        return react_1["default"].createElement(ReadOnlyNickname, { value: this.props.value });
    };
    return Nickname;
}(react_1["default"].PureComponent));
exports.Nickname = Nickname;
var MemoEditableNickname = react_1["default"].memo(EditableNickname);
function EditableNickname(props) {
    var _a = react_1["default"].useState(props.value), value = _a[0], setValue = _a[1];
    var handleChange = function (e) { return setValue(e.target.value); };
    function handleKeyPress(e) {
        if (e.key === 'Enter') {
            e.currentTarget.blur();
        }
    }
    function update() {
        var nick = value;
        if (!value) {
            setValue(window_1.nickname);
            nick = window_1.nickname;
        }
        props.onChange({
            type: 'nickname',
            payload: { nickname: nick }
        });
    }
    return (react_1["default"].createElement("input", { className: "nickname", type: "text", onChange: handleChange, onKeyPress: handleKeyPress, onBlur: update, value: value }));
}
function ReadOnlyNickname(props) {
    return react_1["default"].createElement("span", { className: "nickname" }, props.value);
}
