// peer-calls/lib/client/components/Notifications.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var CSSTransition_1 = __importDefault(require("react-transition-group/CSSTransition"));
var TransitionGroup_1 = __importDefault(require("react-transition-group/TransitionGroup"));
var react_1 = __importDefault(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var transitionTimeout = {
    enter: 200,
    exit: 100
};
var Notification = react_1["default"].memo(function Notification(props) {
    var notification = props.notification;
    return (react_1["default"].createElement("div", { className: classnames_1["default"](notification.type, 'notification') }, notification.message));
});
var Notifications = /** @class */ (function (_super) {
    __extends(Notifications, _super);
    function Notifications() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Notifications.prototype.render = function () {
        var _a = this.props, dismiss = _a.dismiss, notifications = _a.notifications, max = _a.max;
        return (react_1["default"].createElement("div", { className: "notifications", tabIndex: 0 },
            react_1["default"].createElement(TransitionGroup_1["default"], null, Object.keys(notifications).slice(-max).reverse().map(function (id) { return (react_1["default"].createElement(CSSTransition_1["default"], { key: id, classNames: 'fade', timeout: transitionTimeout },
                react_1["default"].createElement(Notification, { notification: notifications[id], dismiss: dismiss, timeout: 10000 }))); }))));
    };
    Notifications.defaultProps = {
        max: 20
    };
    return Notifications;
}(react_1["default"].PureComponent));
exports["default"] = Notifications;
