// peer-calls/lib/client/components/Toolbar.test.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
jest.mock('../window');
var react_1 = __importDefault(require("react"));
var react_dom_1 = __importDefault(require("react-dom"));
var test_utils_1 = __importDefault(require("react-dom/test-utils"));
var constants_1 = require("../constants");
var window_1 = require("../window");
var Toolbar_1 = __importDefault(require("./Toolbar"));
describe('components/Toolbar', function () {
    var ToolbarWrapper = /** @class */ (function (_super) {
        __extends(ToolbarWrapper, _super);
        function ToolbarWrapper() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.state = {
                stream: null,
                desktopStream: null
            };
            return _this;
        }
        ToolbarWrapper.prototype.render = function () {
            return react_1["default"].createElement(Toolbar_1["default"], { chatVisible: this.props.chatVisible, dialState: this.props.dialState, onToggleChat: this.props.onToggleChat, onHangup: this.props.onHangup, onGetDesktopStream: this.props.onGetDesktopStream, onRemoveStream: this.props.onRemoveStream, onSendFile: this.props.onSendFile, messagesCount: this.props.messagesCount, stream: this.state.stream || this.props.stream, desktopStream: this.state.desktopStream || this.props.desktopStream });
        };
        return ToolbarWrapper;
    }(react_1["default"].PureComponent));
    var node;
    var mediaStream;
    var url;
    var onToggleChat;
    var onSendFile;
    var onHangup;
    var onGetDesktopStream;
    var onRemoveStream;
    var desktopStream;
    var dialState;
    function render() {
        return __awaiter(this, void 0, void 0, function () {
            var div, stream;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dialState = constants_1.DIAL_STATE_IN_CALL;
                        mediaStream = new window_1.MediaStream();
                        onToggleChat = jest.fn();
                        onSendFile = jest.fn();
                        onHangup = jest.fn();
                        onGetDesktopStream = jest.fn().mockImplementation(function () { return Promise.resolve(); });
                        onRemoveStream = jest.fn();
                        div = document.createElement('div');
                        stream = {
                            stream: mediaStream,
                            type: constants_1.STREAM_TYPE_CAMERA,
                            url: url
                        };
                        return [4 /*yield*/, new Promise(function (resolve) {
                                react_dom_1["default"].render(react_1["default"].createElement(ToolbarWrapper, { ref: function (instance) { return resolve(instance); }, dialState: dialState, chatVisible: true, onHangup: onHangup, onToggleChat: onToggleChat, onSendFile: onSendFile, messagesCount: 1, stream: stream, desktopStream: desktopStream, onGetDesktopStream: onGetDesktopStream, onRemoveStream: onRemoveStream }), div);
                            })];
                    case 1:
                        _a.sent();
                        node = div.children[0];
                        return [2 /*return*/];
                }
            });
        });
    }
    beforeEach(function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, render()];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    describe('handleChatClick', function () {
        it('toggle chat', function () {
            expect(onToggleChat.mock.calls.length).toBe(0);
            var button = node.querySelector('.chat');
            test_utils_1["default"].Simulate.click(button);
            expect(onToggleChat.mock.calls.length).toBe(1);
        });
    });
    describe('handleMicClick', function () {
        it('toggle mic', function () {
            var button = node.querySelector('.mute-audio');
            test_utils_1["default"].Simulate.click(button);
            expect(button.classList.contains('on')).toBe(true);
        });
    });
    describe('handleCamClick', function () {
        it('toggle cam', function () {
            var button = node.querySelector('.mute-video');
            test_utils_1["default"].Simulate.click(button);
            expect(button.classList.contains('on')).toBe(true);
        });
    });
    describe('handleFullscreenClick', function () {
        it('toggle fullscreen', function () {
            var button = node.querySelector('.fullscreen');
            test_utils_1["default"].Simulate.click(button);
            expect(button.classList.contains('on')).toBe(false);
        });
    });
    describe('handleHangoutClick', function () {
        it('hangout', function () {
            var button = node.querySelector('.hangup');
            test_utils_1["default"].Simulate.click(button);
            expect(window.location.href).toBe('http://localhost/');
        });
    });
    describe('handleSendFile', function () {
        it('triggers input dialog', function () {
            var sendFileButton = node.querySelector('.send-file');
            var click = jest.fn();
            var file = node.querySelector('input[type=file]');
            file.click = click;
            test_utils_1["default"].Simulate.click(sendFileButton);
            expect(click.mock.calls.length).toBe(1);
        });
    });
    describe('handleSelectFiles', function () {
        it('iterates through files and calls onSendFile for each', function () {
            var file = node.querySelector('input[type=file]');
            var files = [{ name: 'first' }];
            test_utils_1["default"].Simulate.change(file, {
                target: {
                    files: files
                }
            });
            expect(onSendFile.mock.calls).toEqual([[files[0]]]);
        });
    });
    describe('onHangup', function () {
        it('calls onHangup callback', function () {
            expect(onHangup.mock.calls.length).toBe(0);
            var hangup = node.querySelector('.hangup');
            expect(hangup).toBeDefined();
            test_utils_1["default"].Simulate.click(hangup);
            expect(onHangup.mock.calls.length).toBe(1);
        });
    });
    describe('desktop sharing', function () {
        it('starts desktop sharing', function () { return __awaiter(void 0, void 0, void 0, function () {
            var shareDesktop;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        shareDesktop = node.querySelector('.stream-desktop');
                        expect(shareDesktop).toBeDefined();
                        test_utils_1["default"].Simulate.click(shareDesktop);
                        return [4 /*yield*/, Promise.resolve()];
                    case 1:
                        _a.sent();
                        expect(onGetDesktopStream.mock.calls.length).toBe(1);
                        return [2 /*return*/];
                }
            });
        }); });
        it('stops desktop sharing', function () { return __awaiter(void 0, void 0, void 0, function () {
            var shareDesktop;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        desktopStream = {
                            stream: new window_1.MediaStream(),
                            type: constants_1.STREAM_TYPE_DESKTOP
                        };
                        return [4 /*yield*/, render()];
                    case 1:
                        _a.sent();
                        shareDesktop = node.querySelector('.stream-desktop');
                        expect(shareDesktop).toBeDefined();
                        test_utils_1["default"].Simulate.click(shareDesktop);
                        return [4 /*yield*/, Promise.resolve()];
                    case 2:
                        _a.sent();
                        expect(onRemoveStream.mock.calls.length).toBe(1);
                        return [2 /*return*/];
                }
            });
        }); });
    });
});
