// peer-calls/lib/client/components/Video.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var react_1 = __importDefault(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var Nickname_1 = require("./Nickname");
var Dropdown_1 = require("./Dropdown");
var Video = /** @class */ (function (_super) {
    __extends(Video, _super);
    function Video() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.videoRef = react_1["default"].createRef();
        _this.handleClick = function (e) {
            _this.props.play();
        };
        _this.handleMinimize = function () {
            _this.props.onMinimizeToggle({
                userId: _this.props.userId,
                streamId: _this.props.stream && _this.props.stream.stream.id
            });
        };
        _this.handleToggleCover = function () {
            var v = _this.videoRef.current;
            if (v) {
                v.style.objectFit = v.style.objectFit ? '' : 'contain';
            }
        };
        return _this;
    }
    Video.prototype.componentDidMount = function () {
        this.componentDidUpdate();
    };
    Video.prototype.componentDidUpdate = function () {
        var stream = this.props.stream;
        var video = this.videoRef.current;
        var mediaStream = stream && stream.stream || null;
        var url = stream && stream.url;
        if ('srcObject' in video) {
            if (video.srcObject !== mediaStream) {
                video.srcObject = mediaStream;
            }
        }
        else if (video.src !== url) {
            video.src = url || '';
        }
    };
    Video.prototype.render = function () {
        var _this = this;
        var _a = this.props, mirrored = _a.mirrored, muted = _a.muted, userId = _a.userId, windowState = _a.windowState;
        var className = classnames_1["default"]('video-container', {
            minimized: windowState === 'minimized',
            mirrored: mirrored
        });
        return (react_1["default"].createElement("div", { className: className },
            react_1["default"].createElement("video", { id: "video-" + userId, autoPlay: true, onClick: this.handleClick, onLoadedMetadata: function () { return _this.props.play(); }, playsInline: true, ref: this.videoRef, muted: muted }),
            react_1["default"].createElement("div", { className: 'video-footer' },
                react_1["default"].createElement(Nickname_1.Nickname, { value: this.props.nickname, onChange: this.props.onChangeNickname, localUser: this.props.localUser }),
                react_1["default"].createElement(Dropdown_1.Dropdown, { label: '☰' },
                    react_1["default"].createElement("li", { className: 'action-minimize', onClick: this.handleMinimize }, "Toggle Minimize"),
                    react_1["default"].createElement("li", { className: 'action-toggle-fit', onClick: this.handleToggleCover }, "Toggle Fit")))));
    };
    Video.defaultProps = {
        muted: false,
        mirrored: false
    };
    return Video;
}(react_1["default"].PureComponent));
exports["default"] = Video;
