// peer-calls/lib/server/configureStores.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
jest.mock('ioredis');
var ioredis_1 = __importDefault(require("ioredis"));
var socket_io_1 = __importDefault(require("socket.io"));
var configureStores_1 = require("./configureStores");
var store_1 = require("./store");
describe('configureStores', function () {
    describe('memory', function () {
        it('should be in memory when no params specified', function () {
            var io = socket_io_1["default"]();
            var stores = configureStores_1.configureStores(io);
            expect(stores.socketIdByUserId).toEqual(jasmine.any(store_1.MemoryStore));
            expect(stores.userIdBySocketId).toEqual(jasmine.any(store_1.MemoryStore));
        });
        it('should be in memory when type="memory"', function () {
            var io = socket_io_1["default"]();
            var stores = configureStores_1.configureStores(io);
            expect(stores.socketIdByUserId).toEqual(jasmine.any(store_1.MemoryStore));
            expect(stores.userIdBySocketId).toEqual(jasmine.any(store_1.MemoryStore));
        });
    });
    describe('redis', function () {
        it('should be redis when type="redis"', function () {
            var io = socket_io_1["default"]();
            var stores = configureStores_1.configureStores(io, {
                type: 'redis',
                host: 'localhost',
                port: 6379,
                prefix: 'peercalls'
            });
            expect(io.adapter().pubClient).toEqual(jasmine.any(ioredis_1["default"]));
            expect(io.adapter().subClient).toEqual(jasmine.any(ioredis_1["default"]));
            expect(stores.socketIdByUserId).toEqual(jasmine.any(store_1.RedisStore));
            expect(stores.userIdBySocketId).toEqual(jasmine.any(store_1.RedisStore));
        });
    });
});
