// peer-calls/lib/server/configureStores.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var debug_1 = __importDefault(require("debug"));
var ioredis_1 = __importDefault(require("ioredis"));
var socket_io_redis_1 = __importDefault(require("socket.io-redis"));
var store_1 = require("./store");
var debug = debug_1["default"]('peercalls');
function configureStores(io, config) {
    if (config === void 0) { config = { type: 'memory' }; }
    switch (config.type) {
        case 'redis':
            debug('Using redis store: %s:%s', config.host, config.port);
            configureRedis(io, config);
            return {
                socketIdByUserId: new store_1.RedisStore(createRedisClient(config), [config.prefix, 'socketIdByUserId'].join(':')),
                userIdBySocketId: new store_1.RedisStore(createRedisClient(config), [config.prefix, 'socketIdByUserId'].join(':'))
            };
        default:
            debug('Using in-memory store');
            return {
                socketIdByUserId: new store_1.MemoryStore(),
                userIdBySocketId: new store_1.MemoryStore()
            };
    }
}
exports.configureStores = configureStores;
function configureRedis(io, config) {
    var pubClient = createRedisClient(config);
    var subClient = createRedisClient(config);
    io.adapter(socket_io_redis_1["default"]({
        key: 'peercalls',
        pubClient: pubClient,
        subClient: subClient
    }));
}
function createRedisClient(config) {
    return new ioredis_1["default"]({
        host: config.host,
        port: config.port
    });
}
