// peer-calls/lib/server/socket.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var events_1 = require("events");
var socket_1 = __importDefault(require("./socket"));
var store_1 = require("./store");
describe('server/socket', function () {
    var socket;
    var io;
    var rooms;
    var socket0 = {
        id: 'socket0'
    };
    var socket1 = {
        id: 'socket1'
    };
    var socket2 = {
        id: 'socket2'
    };
    var emitPromise;
    beforeEach(function () {
        socket = new events_1.EventEmitter();
        socket.id = 'socket0';
        socket.join = jest.fn();
        socket.leave = jest.fn();
        rooms = {};
        var emitResolve;
        emitPromise = new Promise(function (resolve) {
            emitResolve = resolve;
        });
        var socketsByRoom = {
            room1: [socket0.id],
            room2: [socket0.id],
            room3: [socket0.id, socket1.id, socket2.id]
        };
        io = {};
        io["in"] = io.to = jest.fn().mockImplementation(function (room) {
            return (rooms[room] = rooms[room] || {
                emit: jest.fn().mockImplementation(function () { return emitResolve(); }),
                clients: function (callback) {
                    callback(undefined, socketsByRoom[room] || []);
                }
            });
        });
    });
    it('should be a function', function () {
        expect(typeof socket_1["default"]).toBe('function');
    });
    describe('socket events', function () {
        var stores;
        beforeEach(function () {
            stores = {
                userIdBySocketId: new store_1.MemoryStore(),
                socketIdByUserId: new store_1.MemoryStore()
            };
            stores.socketIdByUserId.set('a', socket0.id);
            stores.userIdBySocketId.set(socket0.id, 'a');
            stores.socketIdByUserId.set('b', socket1.id);
            stores.userIdBySocketId.set(socket1.id, 'b');
            stores.socketIdByUserId.set('c', socket2.id);
            stores.userIdBySocketId.set(socket2.id, 'c');
            socket_1["default"](socket, io, stores);
        });
        describe('signal', function () {
            it('should broadcast signal to specific user', function () { return __awaiter(void 0, void 0, void 0, function () {
                var signal;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            signal = { type: 'signal' };
                            socket.emit('signal', { userId: 'b', signal: signal });
                            return [4 /*yield*/, emitPromise];
                        case 1:
                            _a.sent();
                            expect(io.to.mock.calls).toEqual([[socket1.id]]);
                            expect(io.to(socket1.id).emit.mock.calls).toEqual([[
                                    'signal', {
                                        userId: 'a',
                                        signal: signal
                                    },
                                ]]);
                            return [2 /*return*/];
                    }
                });
            }); });
        });
        describe('ready', function () {
            it('never calls socket.leave', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            socket.room = 'room1';
                            socket.emit('ready', {
                                userId: 'a',
                                room: 'room2'
                            });
                            return [4 /*yield*/, emitPromise];
                        case 1:
                            _a.sent();
                            expect(socket.leave.mock.calls).toEqual([]);
                            expect(socket.join.mock.calls).toEqual([['room2']]);
                            return [2 /*return*/];
                    }
                });
            }); });
            it('should call socket.join to room', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            socket.emit('ready', {
                                userId: 'b',
                                room: 'room3'
                            });
                            return [4 /*yield*/, emitPromise];
                        case 1:
                            _a.sent();
                            expect(socket.join.mock.calls).toEqual([['room3']]);
                            return [2 /*return*/];
                    }
                });
            }); });
            it('should emit users', function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            socket.emit('ready', {
                                userId: 'a',
                                room: 'room3'
                            });
                            return [4 /*yield*/, emitPromise
                                // expect(io.to.mock.calls).toEqual([[ 'room3' ]])
                            ];
                        case 1:
                            _a.sent();
                            // expect(io.to.mock.calls).toEqual([[ 'room3' ]])
                            expect(io.to('room3').emit.mock.calls).toEqual([
                                [
                                    'users', {
                                        initiator: 'a',
                                        users: [{
                                                socketId: socket0.id,
                                                userId: 'a'
                                            }, {
                                                socketId: socket1.id,
                                                userId: 'b'
                                            }, {
                                                socketId: socket2.id,
                                                userId: 'c'
                                            }]
                                    },
                                ],
                            ]);
                            return [2 /*return*/];
                    }
                });
            }); });
        });
    });
});
