// peer-calls/lib/server/socket.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var debug_1 = __importDefault(require("debug"));
var debug = debug_1["default"]('peercalls:socket');
function handleSocket(socket, io, stores) {
    var _this = this;
    socket.once('disconnect', function () { return __awaiter(_this, void 0, void 0, function () {
        var userId;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, stores.userIdBySocketId.get(socket.id)];
                case 1:
                    userId = _a.sent();
                    if (!userId) return [3 /*break*/, 3];
                    return [4 /*yield*/, Promise.all([
                            stores.userIdBySocketId.remove(socket.id),
                            stores.socketIdByUserId.remove(userId),
                        ])];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3: return [2 /*return*/];
            }
        });
    }); });
    socket.on('signal', function (payload) { return __awaiter(_this, void 0, void 0, function () {
        var socketId, userId;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, stores.socketIdByUserId.get(payload.userId)];
                case 1:
                    socketId = _a.sent();
                    return [4 /*yield*/, stores.userIdBySocketId.get(socket.id)];
                case 2:
                    userId = _a.sent();
                    if (socketId) {
                        io.to(socketId).emit('signal', {
                            userId: userId,
                            signal: payload.signal
                        });
                    }
                    return [2 /*return*/];
            }
        });
    }); });
    socket.on('ready', function (payload) { return __awaiter(_this, void 0, void 0, function () {
        var userId, room, users;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    userId = payload.userId, room = payload.room;
                    debug('ready: %s, room: %s', userId, room);
                    // no need to leave rooms because there will be only one room for the
                    // duration of the socket connection
                    return [4 /*yield*/, Promise.all([
                            stores.socketIdByUserId.set(userId, socket.id),
                            stores.userIdBySocketId.set(socket.id, userId),
                        ])];
                case 1:
                    // no need to leave rooms because there will be only one room for the
                    // duration of the socket connection
                    _a.sent();
                    socket.join(room);
                    return [4 /*yield*/, getUsers(room)];
                case 2:
                    users = _a.sent();
                    debug('ready: %s, room: %s, users: %o', userId, room, users);
                    io.to(room).emit('users', {
                        initiator: userId,
                        users: users
                    });
                    return [2 /*return*/];
            }
        });
    }); });
    function getUsers(room) {
        return __awaiter(this, void 0, void 0, function () {
            var socketIds, userIds;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, getClientsInRoom(room)];
                    case 1:
                        socketIds = _a.sent();
                        return [4 /*yield*/, stores.userIdBySocketId.getMany(socketIds)];
                    case 2:
                        userIds = _a.sent();
                        return [2 /*return*/, socketIds.map(function (socketId, i) { return ({
                                socketId: socketId,
                                userId: userIds[i]
                            }); })];
                }
            });
        });
    }
    function getClientsInRoom(room) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        io["in"](room).clients(function (err, clients) {
                            if (err) {
                                reject(err);
                            }
                            else {
                                resolve(clients);
                            }
                        });
                    })];
            });
        });
    }
}
exports["default"] = handleSocket;
