// peer-calls/lib/server/turn.test.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
exports.__esModule = true;
var turn = __importStar(require("./turn"));
describe('server/turn', function () {
    describe('getCredentials', function () {
        it('returns username & credential', function () {
            var auth = turn.getCredentials('a', 'b');
            expect(auth).toEqual(jasmine.any(Object));
            expect(auth.username).toEqual(jasmine.any(String));
            expect(auth.credential).toEqual(jasmine.any(String));
        });
    });
    describe('processServers', function () {
        var servers = [{
                url: 'server1',
                urls: 'server1',
                auth: undefined,
                username: 'a',
                credential: 'b'
            }, {
                url: 'server2',
                urls: 'server2',
                username: 'c',
                secret: 'd',
                auth: 'secret'
            }];
        it('does not expose secret', function () {
            var s = turn.processServers(servers);
            expect(s.length).toBe(2);
            expect(s[0]).toBe(servers[0]);
            expect(s[1]).toEqual({
                url: 'server2',
                urls: 'server2',
                username: jasmine.any(String),
                credential: jasmine.any(String)
            });
            expect(s[1].username).toMatch(/^[0-9]+:c$/);
        });
        it('throws error when unknown auth type', function () {
            expect(function () { return turn.processServers([{ auth: 'bla' }]); })
                .toThrowError(/not implemented/);
        });
    });
});
