// peer-calls/lib/server/turn.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var crypto_1 = __importDefault(require("crypto"));
function getCredentials(name, secret) {
    // this credential would be valid for the next 24 hours
    var timestamp = Math.floor(Date.now() / 1000) + 24 * 3600;
    var username = [timestamp, name].join(':');
    var hmac = crypto_1["default"].createHmac('sha1', secret);
    hmac.setEncoding('base64');
    hmac.write(username);
    hmac.end();
    var credential = hmac.read();
    return { username: username, credential: credential };
}
exports.getCredentials = getCredentials;
function getServerConfig(server, cred) {
    return {
        url: server.url,
        urls: server.urls,
        username: cred.username,
        credential: cred.credential
    };
}
function processServers(iceServers) {
    return iceServers.map(function (server) {
        switch (server.auth) {
            case undefined:
                return server;
            case 'secret':
                return getServerConfig(server, getCredentials(server.username, server.secret));
            default:
                throw new Error('Authentication type not implemented: ' +
                    server.auth);
        }
    });
}
exports.processServers = processServers;
