// peer-calls/lib/client/constants.js
"use strict";
exports.__esModule = true;
exports.MINIMIZE_TOGGLE = 'MINIMIZE_TOGGLE';
exports.ALERT = 'ALERT';
exports.ALERT_DISMISS = 'ALERT_DISMISS';
exports.ALERT_CLEAR = 'ALERT_CLEAR';
exports.DIAL = 'DIAL';
exports.DIAL_STATE_HUNG_UP = 'hung-up';
exports.DIAL_STATE_DIALLING = 'dialling';
exports.DIAL_STATE_IN_CALL = 'in-call';
exports.HANG_UP = 'HANG_UP';
exports.INIT = 'INIT';
exports.ME = '_me_';
exports.PEERCALLS = '[PeerCalls]';
exports.NOTIFY = 'NOTIFY';
exports.NOTIFY_DISMISS = 'NOTIFY_DISMISS';
exports.NOTIFY_CLEAR = 'NOTIFY_CLEAR';
exports.MESSAGE_ADD = 'MESSAGE_ADD';
exports.MEDIA_ENUMERATE = 'MEDIA_ENUMERATE';
exports.MEDIA_STREAM = 'MEDIA_STREAM';
exports.MEDIA_VIDEO_CONSTRAINT_SET = 'MEDIA_VIDEO_CONSTRAINT_SET';
exports.MEDIA_AUDIO_CONSTRAINT_SET = 'MEDIA_AUDIO_CONSTRAINT_SET';
exports.MEDIA_PLAY = 'MEDIA_PLAY';
exports.NICKNAME_SET = 'NICKNAME_SET';
exports.PEER_ADD = 'PEER_ADD';
exports.PEER_REMOVE = 'PEER_REMOVE';
exports.PEER_EVENT_ERROR = 'error';
exports.PEER_EVENT_CONNECT = 'connect';
exports.PEER_EVENT_CLOSE = 'close';
exports.PEER_EVENT_SIGNAL = 'signal';
exports.PEER_EVENT_TRACK = 'track';
exports.PEER_EVENT_DATA = 'data';
exports.SOCKET_EVENT_READY = 'ready';
exports.SOCKET_EVENT_SIGNAL = 'signal';
exports.SOCKET_EVENT_USERS = 'users';
exports.STREAM_ADD = 'PEER_STREAM_ADD';
exports.STREAM_REMOVE = 'PEER_STREAM_REMOVE';
exports.STREAM_TRACK_ADD = 'PEER_STREAM_TRACK_ADD';
exports.STREAM_TRACK_REMOVE = 'PEER_STREAM_TRACK_REMOVE';
exports.STREAM_TYPE_CAMERA = 'camera';
exports.STREAM_TYPE_DESKTOP = 'desktop';
