// peer-calls/lib/server/store/store.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var _1 = require("./");
var ioredis_1 = __importDefault(require("ioredis"));
describe('store', function () {
    var redis = new ioredis_1["default"]({
        host: process.env.TEST_REDIS_HOST || 'localhost',
        port: parseInt(process.env.TEST_REDIS_PORT) || 6379,
        enableOfflineQueue: true
    });
    var testCases = [{
            name: _1.MemoryStore.name,
            store: new _1.MemoryStore()
        }, {
            name: _1.RedisStore.name,
            store: new _1.RedisStore(redis, 'peercallstest')
        }];
    afterAll(function () {
        redis.disconnect();
    });
    testCases.forEach(function (_a) {
        var name = _a.name, store = _a.store;
        describe(name, function () {
            afterEach(function () { return __awaiter(void 0, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, Promise.all([
                                store.remove('a'),
                                store.remove('b'),
                            ])];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); });
            describe('set, get, getMany', function () {
                it('sets and retreives value(s)', function () { return __awaiter(void 0, void 0, void 0, function () {
                    var _a, _b, _c, _d, _e;
                    return __generator(this, function (_f) {
                        switch (_f.label) {
                            case 0: return [4 /*yield*/, store.set('a', 'A')];
                            case 1:
                                _f.sent();
                                return [4 /*yield*/, store.set('b', 'B')];
                            case 2:
                                _f.sent();
                                _a = expect;
                                return [4 /*yield*/, store.get('a')];
                            case 3:
                                _a.apply(void 0, [_f.sent()]).toBe('A');
                                _b = expect;
                                return [4 /*yield*/, store.get('b')];
                            case 4:
                                _b.apply(void 0, [_f.sent()]).toBe('B');
                                _c = expect;
                                return [4 /*yield*/, store.remove('b')];
                            case 5:
                                _c.apply(void 0, [_f.sent()]);
                                _d = expect;
                                return [4 /*yield*/, store.get('c')];
                            case 6:
                                _d.apply(void 0, [_f.sent()]).toBe(undefined);
                                _e = expect;
                                return [4 /*yield*/, store.getMany(['a', 'b', 'c'])];
                            case 7:
                                _e.apply(void 0, [_f.sent()])
                                    .toEqual(['A', undefined, undefined]);
                                return [2 /*return*/];
                        }
                    });
                }); });
            });
        });
    });
});
